File: 프로그래머스/스택and큐/다리를지나는트럭.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0;
    queue <int> q;
    queue <int> bridge;
    int sum = 0;
    
    for(int i = 0; i < truck_weights.size(); i++){
        q.push(truck_weights[i]);
    }
    
    for(int i = 0; i < bridge_length; i++){
        bridge.push(0);
    }
    
    while(!q.empty() || !bridge.empty() ){
        sum = sum - bridge.front();
        bridge.pop();
        
        if(sum+q.front()<=weight && !q.empty()){
            sum+=q.front();
            bridge.push(q.front());
            q.pop();   
        }
        else if (!q.empty()){
            bridge.push(0);
        }

        
        answer += 1;
        
    }
    
    return answer;
}

File: 프로그래머스_python/분류별/스택and큐/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    answer = 0
    
    queue = deque()
    time_cost = 0
    weight_cost = 0
    index = 0
    
    for i in range(bridge_length):
        queue.append(0)
    
    
    while index < len(truck_weights):
        
        weight_cost -= queue[0]
        queue.popleft()
        time_cost += 1
        
        if weight_cost + truck_weights[index] <= weight:
            queue.append(truck_weights[index])
            weight_cost += truck_weights[index]
            index += 1
        
        else:
            queue.append(0)
            
    while queue:
        queue.popleft()
        time_cost += 1
        
    answer = time_cost
    return answer
